ppcs/ppc0022-metaprogramming.md

```diff
@@ -0,0 +1,385 @@
+
+#### `add_symbol`
+
+```perl
+$metasymbol = $metapackage->add_symbol($name, $value);
+```
+
+Creates a new named symbol. If a value is passed it must be a reference to an item of the compatible type. If no initialisation is provided for variables, an empty variable is created. If no initialisation is provided for a subroutine then a forward declaration is created which has no body yet.
+
+**Optionally on globs** If globs are also supported, then the `$name` can be expressed without a sigil (or maybe the `*` sigil?) to  create and return a metaglob instead of a metasymbol.
+
+#### `remove_symbol`
+
+```perl
+$metapackage->remove_symbol($name);
+```
+
+Removes a symbol. Nothing is returned.
+
+#### `list_symbols`
+
+```perl
+@metasymbols = $metapackage->list_symbols($filter)
+```
+
+Returns a list of symbols in the given package, in no particular order. (And in particular, we make no guarantees about the order of the results as compared to the order of the source code which declared it).
+
+TODO: Define what the filters look like. Need to be able to select scalar/array/hash/code as well as sub-stashes. Actually, do we need sub-stashes if we have `list_packages`?
+
+**Optionally on globs** If globs are also supported, this raises the fun question of whether the returned list contains the actual globs or only the symbols from within them. Perhaps there should be another method `->list_globs` that yields just globs?
+
+### Methods on Meta-Symbol Objects
+
+Several of the above methods return instances of a meta-symbol object. In practice, any meta-symbol object will represent either a variable (when using the "$@%" sigils), or a subroutine (when using the "&"). Each of these cases is described further below. The following methods are available on both kinds.
+
+#### `basename`
+
+```perl
+$name = $metasymbol->basename;
+```
+
+Returns the final part of the symbol's fully qualified name; the name within the package after the final `::` but **excluding** the leading sigil.
+
+#### `sigil`
+
+```perl
+$sigil = $metasymbol->sigil;
+```
+
+Returns the single-character sigil that leads the symbol's name. This will be one of "$", "@", "%" for variables or "&" for subroutines.
+
+#### `package`
+
+```perl
+$metapackage = $metasymbol->package;
+```
+
+Returns the meta-package object instance representing the package to which this symbol belongs.
+
+#### `name`
+
+```perl
+$name = $metasymbol->name;
+```
+
+Returns the fully-qualified name for this symbol, including its sigil and package name. This would be equivalent to
+
+```perl
+$name = $metasymbol->sigil . join "::", $metasymbol->package->name, $metasymbol->basename;
+```
+
+#### `is_`*\**
+
+```perl
+$bool = $metasymbol->is_scalar;
+$bool = $metasymbol->is_array;
+$bool = $metasymbol->is_hash;
+$bool = $metasymbol->is_subroutine;
+```
+
+Returns true in each of the four cases where the object represents the given type of symbol, or false in the other three cases.
```

Comment:
For completeness, what about `is_filehandle`?

Reply:
> For completeness, what about `is_filehandle`?

Well currently the API doesn't have the shape to permit access to filehandles. But that's an easy extension that could be added later.

The point isn't to guarantee 100% coverage from the start; just to act as a enough of an initial shape to hang more onto later when we need it. Once there's a bunch of "plain boring objects" with "plain boring methods" on them, it's quite easy to imagine what new methods can be added for those things.

Reply:
If it can't be handled by the shape of this API, then the API seems wrong. And the shape seems to be the most important part to get right to begin with.

If symbols are only accessible via name with a sigil, then file handles won't ever be accessible. That probably indicates that the type needs to be able to be specified separately.

Reply:
rewritten with globs as an inherent first-class concept (and hinting that the direct access to variables is a shortcut via those) in 422c342